### PR TITLE
Add plugin for mysqlclient

### PR DESCRIPTION
**Pull request by katelei6**

### CODE REVIEW

**Comment by katelei6** [comment]
@wu-sheng we add mysqlclient plugin,

File: skywalking/config.py [hunk 1]
```diff
@@ -86,6 +86,11 @@
                       '%(asctime)s [%(threadName)s] %(levelname)s %(name)s - %(message)s'  # type: str
 cause_exception_depth = int(os.getenv('SW_AGENT_CAUSE_EXCEPTION_DEPTH') or '5')  # type: int
 
+mysqlclient_host = os.getenv('SW_MYSQLCLIENT_HOST') or '127.0.0.1'
+mysqlclient_port = os.getenv('SW_MYSQLCLIENT_PORT') or '3306'
+mysqlclient_db = os.getenv('SW_MYSQLCLIENT_DB') or ''
```
Comment (kezhenxu94) [inline skywalking/config.py:91]: These should not be configured, they should be detected automatically.

Comment (katelei6) [reply skywalking/config.py:91]: ok，we edit it.

**Comment by katelei6** [comment]
The host and port attributes can be obtained, but there is no function to obtain the db attribute. is it necessary to put the db in the configuration file？

**Comment by wu-sheng** [comment]
Database information for what? The user's database connection should be detected from the target codes.
No one should provide this database information manually.

**Comment by katelei6** [comment]
@wu-sheng  host and port attributes can be obtained, but there is no function to obtain the db attribute. is it necessary to put the db in the configuration file？

**Comment by wu-sheng** [comment]
> host and port attributes can be obtained, but there is no function to obtain the db attribute. is it necessary to put the db in the configuration file？

I am not following. This is just a tracing plugin, everything that user's code involved, could be detected by intercepting the methods one way or another.

File: tests/plugin/data/sw_mysqlclient/provider.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+if __name__ == '__main__':
+    from flask import Flask, jsonify
+    import MySQLdb
+
+    app = Flask(__name__)
+
+    @app.route('/users', methods=['POST', 'GET'])
+    def application():
+        time.sleep(0.5)
+        connection = MySQLdb.connect(host='127.0.0.1', user='root', passwd='123456', db='mysql', charset='utf8mb4')
```
Comment (kezhenxu94) [inline tests/plugin/data/sw_mysqlclient/provider.py:29]: You should intercept this method, there is all kinds of information you need, including `host`, `db`, etc.

File: skywalking/plugins/sw_mysqlclient.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    from MySQLdb.cursors import Cursor
+
+    _execute = Cursor.execute
+
+    def _sw_execute(this: Cursor, query, args=None):
+        peer = f'{this.connection.get_host_info().split()[0]}:{this.connection.port}'
+        context = get_context()
+        with context.new_exit_span(op='Mysql/MysqlClient/execute', peer=peer, component=Component.MysqlClient) as span:
+            span.layer = Layer.Database
+            _execute(this, 'select database()')
```
Comment (kezhenxu94) [inline skywalking/plugins/sw_mysqlclient.py:42]: I'd rather not to execute sql just for the database name.

Did you read this https://github.com/apache/skywalking-python/pull/178#discussion_r793186936 ?

Comment (katelei6) [reply skywalking/plugins/sw_mysqlclient.py]: @kezhenxu94 we change the method to intercept

File: skywalking/config.py (hunk 1 above)
Comment (kezhenxu94) [reply skywalking/config.py:91]: These should be removed now

**Review by kezhenxu94 — COMMENTED** [review]
Can you also fix the failed tests?

**Comment by kezhenxu94** [comment]
Hi @katelei6 , can you please take a look at the failed test cases?

**Comment by katelei6** [comment]
I edit the unit test now

**Comment by katelei6** [comment]
@kezhenxu94  I edit the test, but the comsumer run failed!  I run it successfully in our env, so how can i test it ?

**Comment by katelei6** [comment]
the consumer run failed?

**Comment by kezhenxu94** [comment]
> the consumer run failed?

Let me take a look

**Comment by katelei6** [comment]
@kezhenxu94 Thanks very much！ I am a java developer，python i do not very proficient， so thanks to help me look it!

File: skywalking/plugins/sw_mysqlclient.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#
+
+from skywalking import Layer, Component, config
+from skywalking.trace.context import get_context
+from skywalking.trace.tags import TagDbType, TagDbInstance, TagDbStatement, TagDbSqlParameters
+
+link_vector = ['https://mysqlclient.readthedocs.io/']
+support_matrix = {
+    'mysqlclient': {
+        '>=2.1.0': ['1.0']
```
Comment (kezhenxu94) [inline skywalking/plugins/sw_mysqlclient.py:25]: The `support_matrix` is `Python Version`: ['lib version'], so I think you might want 

```suggestion
        '>=3.6': ['2.1.0']
```

File: skywalking/plugins/sw_mysqlclient.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+            self._self_cur = cur
+
+        def __enter__(self):
+            return ProxyCursor(wrapt.ObjectProxy.__enter__(self))
+
+        def execute(self, query, args=None):
+            peer = f'{self.connection.host}:{self.connection.port}'
+            with get_context().new_exit_span(op='Mysql/MysqlClient/execute', peer=peer,
+                                             component=Component.MysqlClient) as span:
```
Comment (kezhenxu94) [inline skywalking/plugins/sw_mysqlclient.py:55]: `Component.MysqlClient` is not defined, you should also add the component and a mapping in our main repo https://github.com/apache/skywalking/blob/473debfee363b782cfe6b18dcb7ec54da5b29f77/oap-server/server-starter/src/main/resources/component-libraries.yml#L564 
https://github.com/apache/skywalking/blob/473debfee363b782cfe6b18dcb7ec54da5b29f77/oap-server/server-starter/src/main/resources/component-libraries.yml#L649

Comment (katelei6) [reply skywalking/plugins/sw_mysqlclient.py:55]: I add Component MysqlClient in __init__.py，and add in the component-libraries.yml too.

**Comment by kezhenxu94** [comment]
The tests failed because of the wrong MySQL url, you need to use the container name as host.

File: tests/plugin/data/sw_mysqlclient/services/provider.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+if __name__ == '__main__':
+    from flask import Flask, jsonify
+    import MySQLdb
+
+    app = Flask(__name__)
+
+    @app.route('/users', methods=['POST', 'GET'])
+    def application():
+        time.sleep(0.5)
+        connection = MySQLdb.connect(host='127.0.0.1', user='root', passwd='root', db='mysql', charset='utf8mb4')
```
Comment (kezhenxu94) [inline tests/plugin/data/sw_mysqlclient/services/provider.py:29]:
```suggestion
        connection = MySQLdb.connect(host='mysql', user='root', passwd='root', db='mysql', charset='utf8mb4')
```

File: skywalking/plugins/sw_mysqlclient.py (hunk 5 above)
Comment (katelei6) [reply skywalking/plugins/sw_mysqlclient.py:55]: I see you update 127.0.0.1 to mysql

Comment (kezhenxu94) [reply skywalking/plugins/sw_mysqlclient.py:55]: > I see you update 127.0.0.1 to mysql

yes

Comment (katelei6) [reply skywalking/plugins/sw_mysqlclient.py:55]: How about fastapi？we want to use this in our new project ，so we will dev the plugin.  But we find fastapi is different with flask, so can you give us prompt or suggestion to diraction?thanks very much!

Comment (kezhenxu94) [reply skywalking/plugins/sw_mysqlclient.py:55]: > How about fastapi？we want to use this in our new project ，so we will dev the plugin.  But we find fastapi is different with flask, so can you give us prompt or suggestion to diraction?thanks very much!

The idea is the same , find out the key method of the framework that will be invoked when serving an API, and intercept it. What do you mean it's different from Flask?

Comment (katelei6) [reply skywalking/plugins/sw_mysqlclient.py:55]: I look the flask，intercept the method through request to response,but fastapi we has not find now! Thanks，we find it.

**Review by kezhenxu94 — APPROVED** [review]
LGTM, thanks @katelei6

**Review by kezhenxu94 — APPROVED** [review]

**Comment by kezhenxu94** [comment]
@katelei6 remember to update our main repo and UI repo (if needed) as https://github.com/apache/skywalking-python/pull/178#discussion_r794309488 ed

File: skywalking/__init__.py [hunk 7]
```diff
@@ -44,6 +44,7 @@ class Component(Enum):
     Psycopg = 7010
     Celery = 7011
     Falcon = 7012
+    MysqlClient = 7013
```
Comment (wu-sheng) [inline skywalking/__init__.py:47]: 7013 is not registered in the main repo, it is better to do that before merging.

**Comment by katelei6** [comment]
https://github.com/katelei6/skywalking  We add the component and push it. You can help me review it. @wu-sheng  @kezhenxu94

**Comment by wu-sheng** [comment]
@katelei6 This is your fork, you need a pull request to the upstream.

**Comment by katelei6** [comment]
Ok, I do it right now.

**Review by wu-sheng — APPROVED** [review]
Main repo PR, https://github.com/apache/skywalking/pull/8487

**Pull request merged**